Assets/MixedRealityToolkit.SDK/Features/Utilities/StabilizationPlaneModifier.cs

```diff
@@ -0,0 +1,390 @@
+                }
+                return inputSystem;
+            }
+        }
+
+        /// <summary>
+        /// Position of the plane in world space.
+        /// </summary>
+        private Vector3 planePosition;
+
+        /// <summary>
+        /// Current distance of the plane from the user's head. Only used when not using the target override
+        /// or the GazeManager to set the plane's position. 
+        /// </summary>
+        private float currentPlaneDistance = 4.0f;
+
+        /// <summary>
+        /// Tracks the previous position of the target override object. Used if velocity is being tracked.
+        /// </summary>
+        private Vector3 targetOverridePreviousPosition;
+
+        /// <summary>
+        /// Used for representing latest plane drawn as gizmo
+        /// </summary>
+        private StabilizationPlaneOverride debugPlane;
+
+        /// <summary>
+        /// Used for representing the debug mesh
+        /// </summary>
+        private GameObject debugMesh;
+
+        private void Awake()
+        {
+#if UNITY_EDITOR
+            debugMesh = GameObject.CreatePrimitive(PrimitiveType.Quad);
+            debugMesh.gameObject.SetActive(false);
+#endif
+        }
+
+        /// <summary>
+        /// Updates the focus point for every frame after all objects have finished moving.
+        /// </summary>
+        private void LateUpdate()
+        {
+            float deltaTime = UseUnscaledTime
+                ? Time.unscaledDeltaTime
+                : Time.deltaTime;
+
+            switch (Mode)
+            {
+                case StabilizationPlaneMode.Fixed:
+                    ConfigureFixedDistancePlane(deltaTime);
+                    break;
+                case StabilizationPlaneMode.GazeHit:
+                    ConfigureGazeManagerPlane(deltaTime);
+                    break;
+                case StabilizationPlaneMode.PlaneOverride:
+                    ConfigureOverridePlane(deltaTime);
+                    break;
+                case StabilizationPlaneMode.TargetOverride:
+                    ConfigureTransformOverridePlane(deltaTime);
+                    break;
+                case StabilizationPlaneMode.Off:
+                default:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Called by Unity when this script is loaded or a value is changed in the inspector.
+        /// Only called in editor, ensures that the property values always match the corresponding member variables.
+        /// </summary>
+        private void OnValidate()
+        {
+            TrackVelocity = trackVelocity;
+            TargetOverride = targetOverride;
+        }
+
+        /// <summary>
+        /// Gets the origin of the gaze for purposes of placing the stabilization plane
+        /// </summary>
+        private Vector3 GazeOrigin
+        {
+            get
+            {
+                if (InputSystem.GazeProvider.Enabled)
```

Comment:
>InputSystem [](start = 20, length = 11)

Note that it's possible the InputSystem could be null if it hasn't initialized yet - this is a separate monobehavior that initializes separate from the input system itself, so depending on timing it could work out poorly in the first frame or work out just fine.

The simplest thing to avoid this would be to add a InputSystem?.GazeProvider...

Reply:
Same comments for the stuff down below.

---
In reply to: [295401779](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4965#discussion_r295401779) [](ancestors = 295401779)